source/opt/dominator_tree.h

```diff
@@ -0,0 +1,180 @@
+  DominatorTreeNode* GetRoot() {
+    assert(roots_.size() == 1);
+    return *roots_.begin();
+  }
+
+  const DominatorTreeNode* GetRoot() const {
+    assert(roots_.size() == 1);
+    return *roots_.begin();
+  }
+
+  const DominatorTreeNodeList& Roots() const {
+    return roots_;
+  }
+
+  // Dumps the tree in the graphvis dot format into the stream.
+  void DumpTreeAsDot(std::ostream& out_stream) const;
+
+  // Build the (post-)dominator tree for the function |F|
+  // Any existing data will be overwritten
+  void InitializeTree(const ir::Function* F);
+
+  // Check if BasicBlock B is a dominator of BasicBlock A.
+  bool Dominates(const ir::BasicBlock* A, const ir::BasicBlock* B) const;
+
+  // Check if BasicBlock B is a dominator of BasicBlock A. This function uses
+  // the IDs of A and B.
+  bool Dominates(uint32_t A, uint32_t B) const;
+
+  // Check if BasicBlock A strictly dominates B
+  bool StrictlyDominates(const ir::BasicBlock* A,
+                         const ir::BasicBlock* B) const;
+
+  bool StrictlyDominates(uint32_t A, uint32_t B) const;
+
+  // Returns the immediate dominator of basicblock A.
+  ir::BasicBlock* ImmediateDominator(const ir::BasicBlock* A) const;
+
+  // Returns the immediate dominator of basicblock A.
+  ir::BasicBlock* ImmediateDominator(uint32_t A) const;
+
+  // Returns true if BasicBlock A is reachable by this tree. A node would be
+  // unreachable if it cannot be reached by traversal from the start node or for
+  // a postdominator tree, cannot be reached from the exit nodes.
+  inline bool ReachableFromRoots(const ir::BasicBlock* A) const {
+    if (!A) return false;
+    return ReachableFromRoots(A->id());
+  }
+
+  // Same as the above method but takes in the ID of the BasicBlock rather than
+  // the BasicBlock itself.
+  bool ReachableFromRoots(uint32_t A) const;
+
+  // Returns true if this tree is a post dominator tree or not.
+  bool IsPostDominator() const { return postdominator_; }
+
+  // Clean up the tree.
+  void ClearTree() {
+    nodes_.clear();
+    roots_.clear();
+  }
+
+ private:
+  // Adds the BasicBlock to the tree structure if it doesn't already exist.
+  DominatorTreeNode* GetOrInsertNode(ir::BasicBlock* bb);
+
+  // Applies the std::function 'func' to 'node' then applies it to node's
+  // children.
+  void Visit(const DominatorTreeNode* node,
+             std::function<void(const DominatorTreeNode*)> func) const;
+
+  // Wrapper function which gets the list of BasicBlock->DominatingBasicBlock
+  // from the CFA and stores it in the edges parameter.
+  //
+  // The |edges| vector will contain the dominator tree as pairs of nodes.
+  // The first node in the pair is a node in the graph. The second node in the
+  // pair is its immediate dominator.
+  // The root of the tree has him self as immediate dominator.
+  void GetDominatorEdges(
+      const ir::Function* f, ir::BasicBlock* dummy_start_node,
+      std::vector<std::pair<ir::BasicBlock*, ir::BasicBlock*>>& edges);
```

Comment:
Style guide: Only const reference arguments. Please change edges to a pointer to vector.